<!DOCTYPE html>
<html lang="fr">

<head>
    <meta charset="UTF-8">
    <meta http-equiv="X-UA-Compatible" content="IE=edge">
    <meta name="viewport" content="width=device-width, initial-scale=1.0">
    <title>Promesse : des requettes</title>
    <style>
        #result {
            background-color: #dfd;
            width: 80%;
            padding: 1rem;
            display: flex;
            flex-direction: column;
            justify-content: center;
            align-items: center;
        }

        .error {
            color: red;
        }
    </style>
</head>

<body>

    <h1>Promise.all()</h1>

    <p>Les promesses s'affichent une fois la réponse obtenue.</p>
    <p>L'ordre d'affichage est geré.</p>
    <p>Si l'une des promesses échoue, Promise.all échoue immédiatement et utilise la raison de l'échec (que les autres
        promesses aient été résolues ou non).</p>
    <div id="result"></div>
    <script>

        /* function pour durée aléatoire */
        alea = () => Math.random() * 2000

        /* creation et affichage d'un paragraphe dans le body */
        const afficheParagraphe = (text, error = false) => {
            const p = document.createElement('p')
            p.innerHTML = text
            if (error) p.classList.add("error")
            document.querySelector('#result').append(p)
        }

        /* une promesse resolue */
        const promesseOk = function (num) {
            return new Promise((resolve, reject) => {
                resolve(setTimeout((num) => resolve(num), alea(100)))
            }
            )
        }
        /* une promesse rejetée */
        const promomeseFail = function (num) {
            return new Promise((resolve, reject) => {
                reject(`erreur promesse ${num}`)
            }
            )
        }


        Promise.all([
            promesseOk(1),
            Promise.resolve(2),
            promesse = 3
            // promomeseFail(4),
            /* promesse tenue car gestion de l'erreur */
            // promomeseFail(5).catch(error => { return error }),
        ])
            .then(resp => {
                /* la réponse est un tableau */
                afficheParagraphe(`réponse ${resp}`)
                resp.forEach(r => afficheParagraphe(r))
            })
            /* si un seule erreur */
            .catch(err => afficheParagraphe(err, true))


        /* a voir pour les curieux ;) */
        // let promiseArray = [
        //     promesseOk(1),
        //     Promise.resolve(2),
        //     promesse = 3
        //     promomeseFail(4),
        //     // promomeseFail(5).catch(error => { return error }),
        // ]
        // let p = Promise.all(promiseArray)
        // console.log(p)
        // setTimeout(function () {
        //     console.log('La pile est vide');
        //     console.log(p);
        // });

    </script>



</body>

</html>